File: example/flux_example_executor/flux_example_executor/merged_sub.py
#!/usr/bin/env python3
"""Serve a flux channel and a ROS subscription from one spin loop.

ros2 run flux_example_executor merged_sub.py
"""

import flux.ros
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.qos import QoSProfile
from rclpy.qos import ReliabilityPolicy
from sensor_msgs.msg import Image as RosImage
from sensor_msgs_flux.image import Image


class MergedSubscriber(Node):

    TOPIC = "image"
    DEPTH = 8

    def __init__(self):
        super().__init__("flux_merged_sub")
        self.flux_seen = 0
        self.ros_seen = 0
        self.flux_width = 0
        self.ros_width = 0

        self.flux_sub = flux.ros.Subscription(
            self, self.TOPIC, callback=self.on_flux, fingerprint=Image.FINGERPRINT__
        )
        qos = QoSProfile(depth=self.DEPTH, reliability=ReliabilityPolicy.BEST_EFFORT)
        self.ros_sub = self.create_subscription(RosImage, self.TOPIC, self.on_ros, qos)
        self.timer = self.create_timer(1.0, self.report)

    def on_flux(self, view):
        v = Image.View(view)
        self.flux_width = v.width
        self.flux_seen += 1

    def on_ros(self, msg):
        self.ros_width = msg.width
        self.ros_seen += 1

    def report(self):
        self.get_logger().info(
            f"flux {self.flux_seen} (w={self.flux_width})  "
            f"ros {self.ros_seen} (w={self.ros_width})"
        )


def main(args=None):
    rclpy.init(args=args)
    node = MergedSubscriber()
    executor = flux.ros.Executor()
    executor.add_flux(node.flux_sub)
    executor.add_ros_node(node)
    node.get_logger().info(f"merged wait: io_uring={executor.uses_io_uring}")
    try:
        executor.spin()
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        executor.stop()
        executor.close()
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: example/flux_example_executor/src/merged_sub.cpp
#include "flux/ros/executor.hpp"
#include "flux/ros/subscription.hpp"

#include <rclcpp/rclcpp.hpp>

#include "sensor_msgs/flux/image.hpp"
#include <sensor_msgs/msg/image.hpp>

#include <chrono>
#include <cstdint>
#include <memory>

// One thread waits on both transports. The flux channel's futex and the ROS subscription's
// readiness eventfd are armed in the same io_uring, so a merged wait costs one syscall.
class MergedSubscriber : public rclcpp::Node
{
public:
  MergedSubscriber()
  : Node("flux_merged_sub"),
    flux_sub_(*this, kTopic, Image::kFingerprint, [this](const flux::FrameView & f) { on_flux(f); })
  {
    ros_sub_ = create_subscription<sensor_msgs::msg::Image>(
      kTopic, rclcpp::QoS(rclcpp::KeepLast(kDepth)).best_effort(),
      [this](const sensor_msgs::msg::Image::ConstSharedPtr & msg) { on_ros(*msg); });
    timer_ = create_wall_timer(std::chrono::seconds(1), [this]() { report(); });
  }

  flux::ros::Subscription & flux_subscription() { return flux_sub_; }

private:
  using Image = sensor_msgs::flux_msg::Image;

  static constexpr const char * kTopic = "image";
  static constexpr std::size_t kDepth = 8;

  void on_flux(const flux::FrameView & f)
  {
    Image::View v{f};
    const std::uint32_t width = v.width();
    if (v.ok__()) {
      flux_width_ = width;
      ++flux_seen_;
    }
  }

  void on_ros(const sensor_msgs::msg::Image & msg)
  {
    ros_width_ = msg.width;
    ++ros_seen_;
  }

  void report()
  {
    RCLCPP_INFO(
      get_logger(), "flux %lu (w=%u)  ros %lu (w=%u)", flux_seen_, flux_width_, ros_seen_,
      ros_width_);
  }

  flux::ros::Subscription flux_sub_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr ros_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::uint64_t flux_seen_ = 0;
  std::uint64_t ros_seen_ = 0;
  std::uint32_t flux_width_ = 0;
  std::uint32_t ros_width_ = 0;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<MergedSubscriber>();

  flux::ros::Executor ex(4);
  ex.add(node->flux_subscription());
  ex.add_ros_node(node);

  rclcpp::on_shutdown([&ex]() { ex.stop(); });

  RCLCPP_INFO(node->get_logger(), "merged wait: io_uring=%d", ex.uses_io_uring());
  ex.spin();
  rclcpp::shutdown();
  return 0;
}
